nn/nnclient.py

```diff
@@ -0,0 +1,305 @@
+       the Free Software Foundation, either version 3 of the License, or
+       (at your option) any later version.
+       This program is distributed in the hope that it will be useful,
+       but WITHOUT ANY WARRANTY; without even the implied warranty of
+       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+       GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see http://www.gnu.org/licenses/.
+
+
+    This module implements a client that connects to a middleware
+    neural network server (see nnserver/nnserver.py), which in turn
+    connects to a TensorFlow model server.
+
+"""
+
+import json
+import requests
+import logging
+
+from settings import Settings
+import tokenizer
+
+from nn import nntree
+from nn.nntree import ParseResult
+from reynir.binparser import BIN_Token
+from reynir import bintokenizer
+
+
+def index_text(text):
+    """ Segments contiguous (Icelandic) text into paragraphs and sentences
+        and returns:
+            dictionary of sentence indices to sentences
+            dictionary of paragraph index to constituent sentence indices"""
+    text = "[[ " + " ]] [[ ".join(text.split("\n")) + " ]]"
+    tok_stream = bintokenizer.tokenize(text)
+
+    pgs = tokenizer.paragraphs(tok_stream)
+    pg_idx_to_sent_idx = dict()
+    sent_idx_to_sent = dict()
+    curr_sent_idx = 0
+    curr_pg_idx = 0
+
+    for pg in pgs:
+        sent_idxs = []
+        for (idx, sent) in pg:
+            curr_sent = list(filter(BIN_Token.is_understood, sent))
+            curr_sent = " ".join([tok.txt for tok in curr_sent])
+            sent_idxs.append(curr_sent_idx)
+            sent_idx_to_sent[curr_sent_idx] = curr_sent
+            curr_sent_idx += 1
+        pg_idx_to_sent_idx[curr_pg_idx] = sent_idxs
+        curr_pg_idx += 1
+    return pg_idx_to_sent_idx, sent_idx_to_sent
+
+
+class NnClient:
+    """ A client that connects to the HTTP rest interface of
+        a tensorflow model server (using plaintext) """
+
+    port = None
+    host = None
+    verb = None
+
+    @classmethod
+    def request_sentence(cls, text):
+        """ Request neural network output for a single sentence """
+        raise NotImplemented
+
+    @classmethod
+    def request_text(cls, text):
+        """ Request neural network output for contiguous text """
+        raise NotImplemented
+
+    @classmethod
+    def _request(cls, pgs, data=None):
+        """ Send serialized request to remote model server """
+        url = "http://{host}:{port}/{verb}.api".format(
+            host=cls.host, port=cls.port, verb=cls.verb
+        )
+        headers = {"content-type": "application/json"}
+        payload = {"pgs": pgs}
+
+        if data is not None:
+            for (k, v) in data.items():
+                payload[k] = v
+
+        logging.debug(str(payload))
```

Comment:
`if Settings.DEBUG: logging.debug(...)`?

Reply:
Think that's part of logging.debug. Messages are ignored if the logging level is not >= than the configured level.

https://github.com/python/cpython/blob/3.8/Lib/logging/__init__.py#L1421

Reply:
Yes, but Settings.DEBUG is our own configuration setting, drawn from the command line when Greynir is started, or from an environment variable, or from a configuration file. Having a single place to control whether debug features are enabled/disabled is a Good Thing.

Reply:
Fully agree a single place to configure this is good. In general the python logging module has this built in with the `--log=DEBUG` option or a global logging configuration.

This can also be configured universally based on the Greynir settings, i.e. `if Settings.debug: app.logger.setLevel(logging.DEBUG)`, and this seems to be happening in line 272 of main.py.

But even if it is wrapped like `if Settings.DEBUG: logging.debug(...)` the logging level would have to be set to` logging.DEBUG` or the conditional body would have no effect, no?